File: packages/core/cloudwright/icons.py
"""Icon registry for cloud service visualization.

Maps (provider, service) pairs to visual metadata: category, color, shape,
ASCII character. Used by ASCII renderer, D2/Mermaid exporters, and React Flow UI.
"""

from __future__ import annotations

from dataclasses import dataclass

# Provider brand colors
PROVIDER_COLORS = {
    "aws": "#FF9900",
    "gcp": "#4285F4",
    "azure": "#0078D4",
}

# Category accent colors (dark-theme friendly)
CATEGORY_COLORS = {
    "compute": "#10b981",
    "database": "#8b5cf6",
    "storage": "#6366f1",
    "network": "#3b82f6",
    "security": "#ef4444",
    "serverless": "#f59e0b",
    "cache": "#8b5cf6",
    "queue": "#f97316",
    "cdn": "#3b82f6",
    "monitoring": "#06b6d4",
    "ml": "#ec4899",
    "analytics": "#a855f7",
}

# Category -> D2/Mermaid node shape
CATEGORY_SHAPES = {
    "compute": "rectangle",
    "database": "cylinder",
    "storage": "rectangle",
    "network": "stadium",
    "security": "rectangle",
    "serverless": "hexagon",
    "cache": "cylinder",
    "queue": "parallelogram",
    "cdn": "stadium",
    "monitoring": "rectangle",
    "ml": "rectangle",
    "analytics": "cylinder",
}

# Category -> ASCII single-char label
CATEGORY_ASCII = {
    "compute": "S",
    "database": "D",
    "storage": "B",
    "network": "N",
    "security": "X",
    "serverless": "F",
    "cache": "H",
    "queue": "Q",
    "cdn": "C",
    "monitoring": "M",
    "ml": "A",
    "analytics": "R",
}

VALID_SHAPES = {"rectangle", "cylinder", "hexagon", "stadium", "parallelogram"}


@dataclass(frozen=True, slots=True)
class ServiceIcon:
    provider: str
    service: str
    category: str
    label: str
    color: str = ""
    shape: str = "rectangle"
    ascii_char: str = "?"

    def __post_init__(self):
        if self.shape not in VALID_SHAPES:
            raise ValueError(f"Invalid shape {self.shape!r}, must be one of {VALID_SHAPES}")


def _icon(provider: str, service: str, category: str, label: str) -> ServiceIcon:
    """Build a ServiceIcon with category-derived defaults."""
    return ServiceIcon(
        provider=provider,
        service=service,
        category=category,
        label=label,
        color=CATEGORY_COLORS.get(category, "#94a3b8"),
        shape=CATEGORY_SHAPES.get(category, "rectangle"),
        ascii_char=CATEGORY_ASCII.get(category, "?"),
    )


# --- AWS ---
_AWS_ICONS: list[ServiceIcon] = [
    _icon("aws", "ec2", "compute", "EC2"),
    _icon("aws", "ecs", "compute", "ECS"),
    _icon("aws", "eks", "compute", "EKS"),
    _icon("aws", "emr", "compute", "EMR"),
    _icon("aws", "rds", "database", "RDS"),
    _icon("aws", "aurora", "database", "Aurora"),
    _icon("aws", "dynamodb", "database", "DynamoDB"),
    _icon("aws", "redshift", "analytics", "Redshift"),
    _icon("aws", "s3", "storage", "S3"),
    _icon("aws", "ebs", "storage", "EBS"),
    _icon("aws", "ecr", "storage", "ECR"),
    _icon("aws", "alb", "network", "ALB"),
    _icon("aws", "nlb", "network", "NLB"),
    _icon("aws", "route53", "network", "Route 53"),
    _icon("aws", "cloudfront", "cdn", "CloudFront"),
    _icon("aws", "waf", "security", "WAF"),
    _icon("aws", "cognito", "security", "Cognito"),
    _icon("aws", "kms", "security", "KMS"),
    _icon("aws", "cloudtrail", "security", "CloudTrail"),
    _icon("aws", "guardduty", "security", "GuardDuty"),
    _icon("aws", "lambda", "serverless", "Lambda"),
    _icon("aws", "api_gateway", "serverless", "API Gateway"),
    _icon("aws", "step_functions", "serverless", "Step Functions"),
    _icon("aws", "glue", "serverless", "Glue"),
    _icon("aws", "elasticache", "cache", "ElastiCache"),
    _icon("aws", "sqs", "queue", "SQS"),
    _icon("aws", "sns", "queue", "SNS"),
    _icon("aws", "kinesis", "queue", "Kinesis"),
    _icon("aws", "cloudwatch", "monitoring", "CloudWatch"),
    _icon("aws", "codepipeline", "compute", "CodePipeline"),
    _icon("aws", "sagemaker", "ml", "SageMaker"),
]

# --- GCP ---
_GCP_ICONS: list[ServiceIcon] = [
    _icon("gcp", "compute_engine", "compute", "Compute Engine"),
    _icon("gcp", "gke", "compute", "GKE"),
    _icon("gcp", "cloud_sql", "database", "Cloud SQL"),
    _icon("gcp", "bigquery", "analytics", "BigQuery"),
    _icon("gcp", "cloud_storage", "storage", "Cloud Storage"),
    _icon("gcp", "cloud_load_balancing", "network", "Cloud LB"),
    _icon("gcp", "cloud_cdn", "cdn", "Cloud CDN"),
    _icon("gcp", "cloud_functions", "serverless", "Cloud Functions"),
    _icon("gcp", "cloud_run", "serverless", "Cloud Run"),
    _icon("gcp", "memorystore", "cache", "Memorystore"),
    _icon("gcp", "pub_sub", "queue", "Pub/Sub"),
]

# --- Azure ---
_AZURE_ICONS: list[ServiceIcon] = [
    _icon("azure", "virtual_machines", "compute", "Virtual Machines"),
    _icon("azure", "aks", "compute", "AKS"),
    _icon("azure", "azure_sql", "database", "Azure SQL"),
    _icon("azure", "cosmos_db", "database", "Cosmos DB"),
    _icon("azure", "blob_storage", "storage", "Blob Storage"),
    _icon("azure", "app_gateway", "network", "App Gateway"),
    _icon("azure", "azure_functions", "serverless", "Azure Functions"),
    _icon("azure", "azure_cache", "cache", "Azure Cache"),
    _icon("azure", "service_bus", "queue", "Service Bus"),
]

# --- Generic fallbacks ---
_GENERIC_ICONS: list[ServiceIcon] = [
    _icon("generic", "user", "network", "User"),
    _icon("generic", "internet", "network", "Internet"),
    _icon("generic", "firewall", "security", "Firewall"),
    _icon("generic", "load_balancer", "network", "Load Balancer"),
    _icon("generic", "database", "database", "Database"),
    _icon("generic", "cache", "cache", "Cache"),
    _icon("generic", "queue", "queue", "Queue"),
    _icon("generic", "storage", "storage", "Storage"),
    _icon("generic", "compute", "compute", "Compute"),
    _icon("generic", "cdn", "cdn", "CDN"),
]

# Build the lookup registry: (provider, service) -> ServiceIcon
ICON_REGISTRY: dict[tuple[str, str], ServiceIcon] = {}
for _icon_list in (_AWS_ICONS, _GCP_ICONS, _AZURE_ICONS, _GENERIC_ICONS):
    for _si in _icon_list:
        ICON_REGISTRY[(_si.provider, _si.service)] = _si

# Default fallback icon
_DEFAULT_ICON = ServiceIcon(
    provider="generic",
    service="unknown",
    category="compute",
    label="Service",
    color="#94a3b8",
    shape="rectangle",
    ascii_char="?",
)

_D2_ICON_BASE = "https://icons.terrastruct.com"
_D2_PROVIDER_PATHS = {
    "aws": "aws",
    "gcp": "gcp",
    "azure": "azure",
}


def get_icon(provider: str, service: str) -> ServiceIcon | None:
    """Look up icon metadata for a (provider, service) pair. Returns None if unknown."""
    return ICON_REGISTRY.get((provider.lower(), service.lower()))


def get_icon_or_default(provider: str, service: str) -> ServiceIcon:
    """Look up icon metadata, returning a generic fallback if unknown."""
    icon = get_icon(provider, service)
    if icon is not None:
        return icon
    # Try generic fallback by guessing category from service name
    for generic_key in ("compute", "database", "cache", "queue", "storage", "cdn"):
        if generic_key in service.lower():
            generic = ICON_REGISTRY.get(("generic", generic_key))
            if generic:
                return generic
    return _DEFAULT_ICON


def get_category_color(category: str) -> str:
    """Get the accent color for a service category."""
    return CATEGORY_COLORS.get(category, "#94a3b8")


def get_icon_url(provider: str, service: str) -> str:
    """Get a terrastruct CDN icon URL for D2 remote rendering."""
    provider_path = _D2_PROVIDER_PATHS.get(provider.lower(), "aws")
    return f"{_D2_ICON_BASE}/{provider_path}/{service}.svg"

File: packages/web/frontend/src/lib/icons.ts
export const CATEGORY_COLORS: Record<string, string> = {
  compute: "#10b981",
  database: "#8b5cf6",
  storage: "#6366f1",
  network: "#3b82f6",
  security: "#ef4444",
  serverless: "#f59e0b",
  cache: "#8b5cf6",
  queue: "#f97316",
  cdn: "#3b82f6",
  monitoring: "#06b6d4",
  ml: "#ec4899",
  analytics: "#a855f7",
  containers: "#14b8a6",
  streaming: "#f97316",
  orchestration: "#a78bfa",
};

export const SERVICE_CATEGORY: Record<string, string> = {
  // AWS compute
  ec2: "compute", ecs: "compute", eks: "compute", emr: "compute",
  fargate: "compute", codepipeline: "compute", codecommit: "storage",
  codebuild: "compute", dms: "compute", migration_hub: "compute",
  // GCP compute
  compute_engine: "compute", gke: "containers", app_engine: "serverless",
  cloud_build: "compute",
  // Azure compute
  virtual_machines: "compute", aks: "containers", container_apps: "containers",
  azure_devops: "compute", azure_migrate: "compute",

  // Databases
  rds: "database", aurora: "database", dynamodb: "database", cloud_sql: "database",
  azure_sql: "database", cosmos_db: "database", redshift: "database", bigquery: "database",
  firestore: "database", spanner: "database", alloydb: "database",

  // Storage
  s3: "storage", cloud_storage: "storage", blob_storage: "storage", ebs: "storage",
  ecr: "storage", fsx: "storage", efs: "storage", artifact_registry: "storage",

  // Network
  alb: "network", nlb: "network", route53: "network", cloud_load_balancing: "network",
  app_gateway: "network", cloud_dns: "network", direct_connect: "network", vpn: "network",
  azure_lb: "network", azure_dns: "network", cloud_interconnect: "network",
  api_management: "network",

  // CDN
  cloudfront: "cdn", cloud_cdn: "cdn", azure_cdn: "cdn",

  // Security
  waf: "security", cognito: "security", kms: "security", cloudtrail: "security",
  guardduty: "security", shield: "security", security_hub: "security",
  config: "security", inspector: "security", cloud_armor: "security",
  firebase_auth: "security", azure_waf: "security", azure_ad: "security",
  azure_firewall: "security", azure_sentinel: "security", azure_policy: "security",

  // Serverless
  lambda: "serverless", api_gateway: "serverless", cloud_functions: "serverless",
  cloud_run: "serverless", azure_functions: "serverless", step_functions: "serverless",
  glue: "serverless", app_service: "serverless",

  // Cache
  elasticache: "cache", memorystore: "cache", azure_cache: "cache",

  // Queue
  sqs: "queue", sns: "queue", pub_sub: "queue", service_bus: "queue", kinesis: "queue",
  eventbridge: "queue",

  // Monitoring
  cloudwatch: "monitoring", cloud_logging: "monitoring", azure_monitor: "monitoring",

  // ML
  sagemaker: "ml", vertex_ai: "ml", azure_ml: "ml",

  // Analytics
  athena: "analytics", dataproc: "analytics", data_factory: "analytics", synapse: "analytics",

  // Streaming
  dataflow: "streaming", event_hubs: "streaming",

  // Orchestration
  cloud_composer: "orchestration", logic_apps: "orchestration",
};

// SVG path data per category (viewBox 0 0 24 24, stroke-based)
export const CATEGORY_ICONS: Record<string, string> = {
  compute: "M5 12H3l9-9 9 9h-2M5 12v7a2 2 0 002 2h10a2 2 0 002-2v-7",
  database: "M12 2C6.48 2 2 4.24 2 7v10c0 2.76 4.48 5 10 5s10-2.24 10-5V7c0-2.76-4.48-5-10-5zM2 12c0 2.76 4.48 5 10 5s10-2.24 10-5",
  storage: "M20 7H4a1 1 0 00-1 1v8a1 1 0 001 1h16a1 1 0 001-1V8a1 1 0 00-1-1zM4 12h16",
  network: "M12 2a10 10 0 100 20 10 10 0 000-20zm0 0a14.5 14.5 0 014 10 14.5 14.5 0 01-4 10 14.5 14.5 0 01-4-10A14.5 14.5 0 0112 2zM2 12h20",
  security: "M12 2l7 4v5c0 5.25-3.5 10.74-7 12-3.5-1.26-7-6.75-7-12V6l7-4z",
  serverless: "M13 2L3 14h9l-1 8 10-12h-9l1-8z",
  cache: "M4 4h16v4H4zM4 10h16v4H4zM4 16h16v4H4z",
  queue: "M4 6h16M4 12h16M4 18h16",
  cdn: "M12 2a10 10 0 100 20 10 10 0 000-20zm-1 17.93A8 8 0 013 12a8 8 0 018-7.93M12 2v20M2 12h20M4.22 7h15.56M4.22 17h15.56",
  monitoring: "M3 3v18h18M7 16l4-8 4 4 4-6",
  ml: "M12 2a4 4 0 014 4c0 1.95-1.4 3.57-3.24 3.9L12 14l-.76-4.1A4 4 0 018 6a4 4 0 014-4zM8 14h8M6 18h12M9 22h6",
  analytics: "M18 20V10M12 20V4M6 20v-6",
  containers: "M21 16V8a2 2 0 00-1-1.73l-7-4a2 2 0 00-2 0l-7 4A2 2 0 003 8v8a2 2 0 001 1.73l7 4a2 2 0 002 0l7-4A2 2 0 0021 16zM3.27 6.96L12 12l8.73-5.04M12 22.08V12",
  streaming: "M2 12c2-3 4-3 6 0s4 3 6 0 4-3 6 0 4 3 6 0",
  orchestration: "M12 2v4M12 18v4M4.93 4.93l2.83 2.83M16.24 16.24l2.83 2.83M2 12h4M18 12h4M4.93 19.07l2.83-2.83M16.24 7.76l2.83-2.83",
};

export function getServiceCategory(service: string): string {
  return SERVICE_CATEGORY[service] || "compute";
}

export function getCategoryColor(category: string): string {
  return CATEGORY_COLORS[category] || "#94a3b8";
}

export function getCategoryIconPath(category: string): string {
  return CATEGORY_ICONS[category] || CATEGORY_ICONS["compute"];
}
